File: promptodroid/app/src/main/java/xyz/cybersapien/promptodroid/ui/widget/PromptsListService.java
package xyz.cybersapien.promptodroid.ui.widget;

import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.concurrent.CountDownLatch;

import xyz.cybersapien.promptodroid.R;
import xyz.cybersapien.promptodroid.data.DataStore;
import xyz.cybersapien.promptodroid.data.model.PromptStory;
import xyz.cybersapien.promptodroid.data.model.User;
import xyz.cybersapien.promptodroid.utils.ConstantsKt;
import xyz.cybersapien.promptodroid.utils.UtilitiesKt;


/**
 * Created by ogcybersapien on 5/1/18.
 */

public class PromptsListService extends RemoteViewsService {

    @Override
    public RemoteViewsFactory onGetViewFactory(Intent intent) {
        return new PromptsListViewFactory(getApplicationContext(), intent);
    }

    static class PromptsListViewFactory implements RemoteViewsFactory {

        private Context context;
        private ArrayList<PromptStory> promptStories;
        private User currentUser;
        private CountDownLatch countDownLatch;


        public PromptsListViewFactory(Context context, Intent intent) {
            this.context = context;
            FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
            currentUser = new User(user);
            promptStories = new ArrayList<>();
        }

        @Override
        public void onCreate() {
        }

        @Override
        public int getCount() {
            return (promptStories == null) ? 0 : promptStories.size();
        }

        @Override
        public RemoteViews getViewAt(int position) {
            RemoteViews view = new RemoteViews(context.getPackageName(), R.layout.widget_item_prompt_story);
            PromptStory story = promptStories.get(position);
            view.setTextViewText(R.id.prompts_title_text_view, story.getStoryTitle());
            view.setTextViewText(R.id.prompts_date_text_view,
                    UtilitiesKt.getFormattedDate(story.getDate())
            );
            view.setTextViewText(R.id.words_count_text_view,
                    UtilitiesKt.getWordCount(context, story.getStoryDetail())
            );
            return view;
        }

        @Override
        public int getViewTypeCount() {
            return 1;
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public boolean hasStableIds() {
            return false;
        }

        @Override
        public RemoteViews getLoadingView() {
            return null;
        }

        @Override
        public void onDataSetChanged() {
            // Do Nothing
            countDownLatch = new CountDownLatch(1);
            getData();
            try {
                countDownLatch.await();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        public void getData() {
            DataStore.getInstance().getUserDataReference(currentUser)
                    .addListenerForSingleValueEvent(new ValueEventListener() {
                        @Override
                        public void onDataChange(DataSnapshot dataSnapshot) {
                            Iterator<DataSnapshot> snapshotIterator = dataSnapshot
                                    .child(ConstantsKt.PROMPT_KEY).getChildren().iterator();
                            ArrayList<PromptStory> stories = new ArrayList<>();
                            for (; snapshotIterator.hasNext(); ) {
                                DataSnapshot currentData = snapshotIterator.next();
                                PromptStory story = currentData.getValue(PromptStory.class);
                                if (story != null) {
                                    stories.add(story);
                                }
                            }
                            if (countDownLatch.getCount() != 0) {
                                promptStories.clear();
                                promptStories.addAll(stories);
                                countDownLatch.countDown();
                            }
                        }

                        @Override
                        public void onCancelled(DatabaseError databaseError) {
                            // Do Nothing
                        }
                    });
        }

        @Override
        public void onDestroy() {
            // Do Nothing
        }
    }
}

File: promptodroid/app/src/main/java/xyz/cybersapien/promptodroid/ui/widget/PromptsListService.kt
package xyz.cybersapien.promptodroid.ui.widget

import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import android.widget.RemoteViewsService
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import xyz.cybersapien.promptodroid.R
import xyz.cybersapien.promptodroid.data.DataStore
import xyz.cybersapien.promptodroid.data.model.PromptStory
import xyz.cybersapien.promptodroid.data.model.User
import xyz.cybersapien.promptodroid.utils.PROMPT_KEY
import xyz.cybersapien.promptodroid.utils.getFormattedDate
import xyz.cybersapien.promptodroid.utils.getWordCount
import java.util.concurrent.CountDownLatch

/**
 * Created by ogcybersapien on 16/1/18.
 */

class PromptsListService : RemoteViewsService() {

    override fun onGetViewFactory(intent: Intent?): RemoteViewsFactory {
        return PromptsListViewFactory(applicationContext)
    }

    internal class PromptsListViewFactory(
            private val context: Context
    ) : RemoteViewsFactory {

        private var promptStories = ArrayList<PromptStory>()
        private var currentUser: User
        private lateinit var countDownLatch: CountDownLatch

        init {
            val fireUser = FirebaseAuth.getInstance().currentUser!!
            currentUser = User(fireUser)
        }

        override fun onCreate() {

        }

        override fun getLoadingView(): RemoteViews? = null

        override fun getItemId(position: Int): Long = position.toLong()

        override fun onDataSetChanged() {
            countDownLatch = CountDownLatch(1)
            getData()
            try {
                countDownLatch.await()
            } catch (error: InterruptedException) {
                error.printStackTrace()
            }
        }

        private fun getData() {
            DataStore.getUserDataReference(currentUser)
                    .addListenerForSingleValueEvent(object : ValueEventListener {
                        override fun onCancelled(p0: DatabaseError?) {}

                        override fun onDataChange(dataSnapshot: DataSnapshot) {
                            val snapShotIterator =
                                    dataSnapshot.child(PROMPT_KEY).children.iterator()
                            val stories = ArrayList<PromptStory>()
                            while (snapShotIterator.hasNext()) {
                                val currentData = snapShotIterator.next()
                                val story = currentData.getValue(PromptStory::class.java)
                                if (story != null) {
                                    stories.add(story)
                                }
                            }
                            if (countDownLatch.count != 0L) {
                                promptStories.clear()
                                promptStories.addAll(stories)
                                countDownLatch.countDown()
                            }
                        }
                    })
        }

        override fun hasStableIds() = false

        override fun getViewAt(position: Int): RemoteViews {
            val views = RemoteViews(context.packageName, R.layout.widget_item_prompt_story)
            val story = promptStories[position]
            views.setTextViewText(R.id.prompts_title_text_view, story.storyTitle)
            views.setTextViewText(R.id.prompts_date_text_view, getFormattedDate(story.date))
            views.setTextViewText(R.id.words_count_text_view, getWordCount(context, story.storyDetail))
            return views
        }

        override fun getCount(): Int = promptStories.size

        override fun getViewTypeCount(): Int = 1

        override fun onDestroy() = Unit

    }
}